lib/datadog/tracing/contrib/rack/header_tagging.rb

```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
+
+module Datadog
+  module Tracing
+    module Contrib
+      module Rack
+        # Matches Rack-style headers with a matcher and sets matching headers into a span.
+        module HeaderTagging
+          def self.tag_request_headers(span, env, configuration)
+            # Wrap env in a case-insensitive Rack-style accessor.
+            headers = env.is_a?(Header::RequestHeaderCollection) ? env : Header::RequestHeaderCollection.new(env)
+
+            # Use global DD_TRACE_HEADER_TAGS if integration-level configuration is not provided
+            tags = if configuration.using_default?(:headers) && Datadog.configuration.tracing.header_tags.configured?
+                     Datadog.configuration.tracing.header_tags.request_tags(headers)
+                   else
+                     whitelist = configuration[:headers][:request] || []
+                     whitelist.each_with_object({}) do |header, result|
+                       header_value = headers.get(header)
+                       unless header_value.nil?
+                         header_tag = Tracing::Metadata::Ext::HTTP::RequestHeaders.to_tag(header)
+                         result[header_tag] = header_value
+                       end
+                     end
+                   end
+
+            span.set_tags(tags)
+          end
+
+          def self.tag_response_headers(span, headers, configuration)
+            # Use global DD_TRACE_HEADER_TAGS if integration-level configuration is not provided
+            tags = if configuration.using_default?(:headers) && Datadog.configuration.tracing.header_tags.configured?
+                     Datadog.configuration.tracing.header_tags.response_tags(
+                       Core::Utils::Hash::CaseInsensitiveWrapper.new(headers)
+                     )
+                   else
+                     whitelist = configuration[:headers][:response] || []
+                     whitelist.each_with_object({}) do |header, result|
+                       if headers.key?(header)
+                         result[Tracing::Metadata::Ext::HTTP::ResponseHeaders.to_tag(header)] = headers[header]
+                       else
+                         # Try a case-insensitive lookup
+                         uppercased_header = header.to_s.upcase
+                         matching_header = headers.keys.find { |h| h.upcase == uppercased_header }
+                         if matching_header
+                           result[Tracing::Metadata::Ext::HTTP::ResponseHeaders.to_tag(header)] = headers[matching_header]
+                         end
```

Comment:
Could we use the `Core::Utils::Hash::CaseInsensitiveWrapper` here? I mostly as, because we add a dedicated module for case insensitive look up but we are doing a hand roll one in the else branch.